cmd/internal/planetscale_edge_database.go

```diff
@@ -185,59 +186,69 @@ func (p PlanetScaleEdgeDatabase) Read(ctx context.Context, w io.Writer, ps Plane
 	}
 
 	currentPosition := lastKnownPosition
+	if currentPosition.Position == "" || currentPosition.LastKnownPk != nil {
+		syncMode = "full"
+	} else {
+		syncMode = "incremental"
+	}
+
 	table := s.Stream
-	readDuration := 1 * time.Minute
+	readDuration := 5 * time.Minute
+	maxRetries := ps.MaxRetries
+
 	preamble := fmt.Sprintf("[%v:%v:%v shard : %v] ", table.Namespace, TabletTypeToString(tabletType), table.Name, currentPosition.Shard)
 
-	for {
-		p.Logger.Log(LOGLEVEL_INFO, preamble+"Peeking to see if there's any new rows")
-		latestCursorPosition, lcErr := p.getLatestCursorPosition(ctx, currentPosition.Shard, currentPosition.Keyspace, table, ps, tabletType)
-		if lcErr != nil {
-			p.Logger.Log(LOGLEVEL_ERROR, preamble+fmt.Sprintf("Error fetching latest cursor position: %+v", lcErr))
-			return currentSerializedCursor, errors.Wrap(err, "Unable to get latest cursor position")
-		}
-		if latestCursorPosition == "" {
-			p.Logger.Log(LOGLEVEL_ERROR, preamble+fmt.Sprintf("Error fetching latest cursor position, was empty string: %+v", latestCursorPosition))
-			return currentSerializedCursor, errors.Wrap(err, "Unable to get latest cursor position")
-		}
+	p.Logger.Log(LOGLEVEL_INFO, preamble+"Peeking to see if there's any new GTIDs")
+	stopPosition, lcErr := p.getStopCursorPosition(ctx, currentPosition.Shard, currentPosition.Keyspace, table, ps, tabletType)
+	if lcErr != nil {
+		p.Logger.Log(LOGLEVEL_ERROR, preamble+fmt.Sprintf("Error fetching latest cursor position: %+v", lcErr))
+		return currentSerializedCursor, errors.Wrap(err, "Unable to get latest cursor position")
+	}
+	if stopPosition == "" {
+		p.Logger.Log(LOGLEVEL_ERROR, preamble+fmt.Sprintf("Error fetching latest cursor position, was empty string: %+v", stopPosition))
+		return currentSerializedCursor, errors.Wrap(err, "Unable to get latest cursor position")
+	}
```

Comment:
I pulled:
- Fetching the desired stop position
- Validating the desired stop position

Out of the for loop so we aren't "advancing the desired stop position" on any kind of error. This helps the sync finish faster.